autopush-common/src/endpoint.rs

```diff
@@ -0,0 +1,37 @@
+use crate::errors::{Result, ResultExt};
+use fernet::MultiFernet;
+use openssl::hash;
+use url::Url;
+use uuid::Uuid;
+
+/// Create an v1 or v2 WebPush endpoint from the identifiers
+///
+/// Both endpoints use bytes instead of hex to reduce ID length.
+//  v1 is the uaid + chid
+//  v2 is the uaid + chid + sha256(key).bytes
+pub fn make_endpoint(
+    uaid: &Uuid,
+    chid: &Uuid,
+    key: Option<&str>,
+    endpoint_url: &str,
+    fernet: &MultiFernet,
+) -> Result<String> {
+    let root = Url::parse(endpoint_url)
+        .and_then(|root| root.join("/wpush/"))
+        .chain_err(|| "Invalid endpoint_url")?;
+    let mut base = hex::decode(uaid.to_simple().to_string()).chain_err(|| "Error decoding")?;
+    base.extend(hex::decode(chid.to_simple().to_string()).chain_err(|| "Error decoding")?);
+
+    if let Some(k) = key {
```

Comment:
So, part of me wonders if we should create a NONCE key when there's none specified. This would ensure that the new URLs have a consistent length and be a little less subject to potential scrutiny.

Not saying we should do that here, but just thinking about it a bit more with the new system we're putting out.

Reply:
v2 endpoints require VAPID, at minimum to validate the public key hash in the endpoint URL. This would require changes to the clients who don't currently provide a key, because they probably also don't send VAPID.

Reply:
Right, but we'd use the nonce for `v1` endpoints that don't include the key. That shouldn't require any modification of the client.